stack/core/src/main/java/org/apache/usergrid/corepersistence/asyncevents/AmazonAsyncEventService.java

```diff
@@ -0,0 +1,425 @@
+                                   final IndexService indexService,
+                                   final EntityCollectionManagerFactory entityCollectionManagerFactory,
+                                   final RxTaskScheduler rxTaskScheduler) {
+
+        this.indexService = indexService;
+        this.entityCollectionManagerFactory = entityCollectionManagerFactory;
+        this.rxTaskScheduler = rxTaskScheduler;
+
+        final QueueScope queueScope = new QueueScopeImpl(QUEUE_NAME);
+        this.queue = queueManagerFactory.getQueueManager(queueScope);
+        this.indexProcessorFig = indexProcessorFig;
+
+        this.writeTimer = metricsFactory.getTimer(AmazonAsyncEventService.class, "write");
+        this.readTimer = metricsFactory.getTimer(AmazonAsyncEventService.class, "read");
+        this.messageProcessingTimer = metricsFactory.getTimer(AmazonAsyncEventService.class, "message.processing");
+        this.indexErrorCounter = metricsFactory.getCounter(AmazonAsyncEventService.class, "error");
+
+
+        //wire up the gauge of inflight message
+        metricsFactory.addGauge(AmazonAsyncEventService.class, "inflight.meter", new Gauge<Long>() {
+            @Override
+            public Long getValue() {
+                return inFlight.longValue();
+            }
+        });
+
+        start();
+    }
+
+
+    /**
+     * Offer the EntityIdScope to SQS
+     */
+    private void offer(final Object operation) {
+        final Timer.Context timer = this.writeTimer.time();
+
+        try {
+            //signal to SQS
+            this.queue.sendMessage(operation);
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to queue message", e);
+        } finally {
+            timer.stop();
+        }
+    }
+
+
+    /**
+     * Take message from SQS
+     */
+    public List<QueueMessage> take() {
+
+        //SQS doesn't support more than 10
+        final Timer.Context timer = this.readTimer.time();
+
+        try {
+            return queue.getMessages(MAX_TAKE,
+                indexProcessorFig.getIndexQueueVisibilityTimeout(),
+                indexProcessorFig.getIndexQueueTimeout(),
+                AsyncEvent.class);
+        }
+        //stop our timer
+        finally {
+            timer.stop();
+        }
+    }
+
+
+    /**
+     * Ack message in SQS
+     */
+    public void ack(final List<QueueMessage> messages) {
+
+        /**
+         * No op
+         */
+        if (messages.size() == 0) {
+            return;
+        }
+
+        queue.commitMessages(messages);
+    }
+
+    /**
+     * Ack message in SQS
+     */
+    public void ack(final QueueMessage message) {
+
+        queue.commitMessage(message);
+    }
+
+    private void handleMessages(final List<QueueMessage> messages) {
+
+        if (logger.isDebugEnabled()) logger.debug("handleMessages with {} message", messages.size());
+
+        for (QueueMessage message : messages) {
```

Comment:
maybe should change to Observable.from(messages).map(message -> {
//choose case
})